homework/src/index.js

```diff
@@ -1,47 +1,110 @@
+
+  const root = document.getElementById("root");
 
-{
   function fetchJSON(url, cb) {
     const xhr = new XMLHttpRequest();
     xhr.open('GET', url);
     xhr.responseType = 'json';
     xhr.onload = () => {
-      if (xhr.status < 400) {
+      if (xhr.status === 200) {
         cb(null, xhr.response);
       } else {
-        cb(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
+        cb({ text: `Error: ${xhr.status} - ${xhr.statusText}`, id: "id", val: "error" });
       }
     };
-    xhr.onerror = () => cb(new Error('Network request failed'));
+    xhr.onerror = () => {
+      cb({ text: "Network request failed", id: "id", val: "error" });
+    };
     xhr.send();
   }
 
-  function createAndAppend(name, parent, options = {}) {
+  function createEl(name, parent, obj) {
     const elem = document.createElement(name);
     parent.appendChild(elem);
-    Object.keys(options).forEach((key) => {
-      const value = options[key];
-      if (key === 'text') {
-        elem.innerText = value;
-      } else {
-        elem.setAttribute(key, value);
-      }
-    });
+    if (obj) {
+      if (obj.text) { elem.innerHTML = obj.text; }
+      if (obj.id) { elem.setAttribute(obj.id, obj.val); }
+    }
     return elem;
   }
 
-  function main(url) {
-    fetchJSON(url, (err, data) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
+  fetchJSON(url, (err, data) => {
+    if (err === null) {
+      const header = createEl("header", root);
+      createEl("h3", header, { text: "HYF Repositories" });
+      const select = createEl("select", header);
+      let names = [];
+      data.forEach(el => {
+        let name = el.name.charAt(0).toUpperCase() + el.name.slice(1);
+        names.push(name);
+      });
+      names.sort();
+      for (let i = 0; i < names.length; i++) {
+        let optn = createEl("option", select, { text: names[i] });
+        optn.value = i;
+      }
+      repos(data, select);
+    } else {
+      createEl("div", root, err);
+    }
+  });
+
+  function repos(data, select) {
+    let article = createEl("article", root);
+
+    select.addEventListener("change", onSelect);
+    function onSelect() {
+      let v1 = select.options[select.selectedIndex].text.toUpperCase();
+      for (let i = 0; i < data.length; i++) {
+        let v2 = data[i].name.toUpperCase();
+        if (v1 === v2) {
+          article.innerHTML = "";
+          repoDetails(data[i], article);
+          contributors(data[i].contributors_url, article);
+        }
+      }
+    }
+    onSelect();
+
+  }
+
+  function repoDetails(details, parent) {
+    let table = createEl("table", parent);
+    let tr = createEl("tr", table);
+    createEl("th", tr, { text: "Repository:" });
+    let td = createEl("td", tr);
+    let a = createEl("a", td, { text: details.name, id: "href", val: details.html_url });
+    a.setAttribute("target", "_blank");
```

Comment:
The original `createAndAppend()` function could have set this `target` attribute.